pipeline/transform.py

```diff
@@ -0,0 +1,299 @@
+                        return convert_date_format(release_date)
+
+            logging.info("No valid release date found for %s", full_url)
+
+        else:
+            logging.warning("Failed to connect to %s, Status Code: %s",
+                            full_url, response.status_code)
+
+    except requests.exceptions.RequestException as e:
+        logging.error("Request failed for %s: %s", artist_url, e)
+
+    except Exception as e:
+        logging.error("An unexpected error occurred: %s", e)
+
+    return ""
+
+
+def extend_sales_from_df(sales_df: pd.DataFrame) -> None:
+    """
+    Coverts the "sale_date" column into datetime format 
+    and extends the given dataframe with month and year
+    of sales.
+    Does nothing if a "sale_date" column doesn't exist.
+    """
+
+    if "sale_date" in sales_df.columns:
+        sales_df["sale_date"] = pd.to_datetime(
+            sales_df["sale_date"], errors="coerce")
+
+        if sales_df["sale_date"].isna().any():
+            logging.warning(
+                "Some 'sale_date' entries could not be converted to datetime.")
+
+    sales_df["sale_year"] = sales_df["sale_date"].dt.year
+    sales_df["sale_month"] = sales_df["sale_date"].dt.month
+
+
+def replace_blank_album_titles(sales_df: pd.DataFrame) -> None:
+    """
+    Replaces "None" entries in the "album_title" column
+    with the name provided in "item_description"
+    if the item is an album.
+    """
+
+    sales_df.loc[(sales_df["item_type"] == "a") &
+                 (sales_df["album_title"].isna() |
+                 (sales_df["album_title"] == "None")),
+                 "album_title"] = sales_df["item_description"]
+
+
+def fill_out_album_and_track(sales_df: pd.DataFrame) -> None:
+    """
+    Replaces the abbreviations for album and track
+    with the full "Album" or "Track" in the dataframe.
+    Does nothing if a "item_type" column doesn't exist.
+    """
+    if "item_type" in sales_df.columns:
+        sales_df["item_type"] = sales_df["item_type"].map(
+            {"a": "album", "t": "track"})
+
+
+def create_sales_dataframe(sales_info: list[dict]) -> pd.DataFrame:
+    """
+    Returns the given sales information as a dataframe.
+    """
+    sales_df = pd.json_normalize(sales_info)
+
+    return sales_df
+
+
+def clean_sales_dataframe(sales_df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Cleans and formats the sales dataframe.
+    """
+
+    extend_sales_from_df(sales_df)
+    replace_blank_album_titles(sales_df)
+    fill_out_album_and_track(sales_df)
+
+    return sales_df
+
+
+def main_transform(sales_data: dict) -> pd.DataFrame:
+    """
+    Fully transforms given sales data into a dataframe
+    and returns it. 
+    Also outputs it as a .csv file.
+    """
+    config_log()
+    sales_info = get_sale_information(sales_data)
+    sales_df = create_sales_dataframe(sales_info)
+    cleaned_sales_df = clean_sales_dataframe(sales_df)
+    cleaned_sales_df.to_csv("MUSIC_DATA.csv", index=None)
```

Comment:
Although this is good for testing, probably don't want it in the prod build

Reply:
If this iteration is approved, I'll remove the .csv output and the file from the push.